kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py

```diff
@@ -0,0 +1,161 @@
+        analysis.FinalizeSolutionStep()
+        analysis.OutputSolutionStep()
+
+def SPRAdaptativeRemeshingRunSolutionLoop(analysis, with_contact = False):
+    """This function executes the solution loop of the AnalysisStage for cases where remeshing may be considered with SPR convergence criteria
+
+        Keyword arguments:
+        analysis The AnalysisStage to be reinitialized
+    """
+
+    # Remeshing adaptively
+    solver = analysis._GetSolver()
+    computing_model_part = solver.GetComputingModelPart()
+    root_model_part = computing_model_part.GetRootModelPart()
+    metric_process = solver.get_metric_process()
+    remeshing_process = solver.get_remeshing_process()
+    convergence_criteria = solver.get_convergence_criterion()
+    builder_and_solver = solver.get_builder_and_solver()
+    mechanical_solution_strategy = solver.get_mechanical_solution_strategy()
+
+    while analysis.KeepAdvancingSolutionLoop():
+        analysis.time = solver.AdvanceInTime(analysis.time)
+        non_linear_iteration = 1
+        while non_linear_iteration <= analysis.non_linear_iterations:
+            if root_model_part.Is(KratosMultiphysics.MODIFIED):
+                ReInitializeSolver(analysis)
+            if non_linear_iteration == 1 or root_model_part.Is(KratosMultiphysics.MODIFIED):
+                analysis.InitializeSolutionStep()
+                solver.Predict()
+                root_model_part.Set(KratosMultiphysics.MODIFIED, False)
+                analysis.is_printing_rank = False
+            computing_model_part.ProcessInfo.SetValue(KratosMultiphysics.NL_ITERATION_NUMBER, non_linear_iteration)
+            is_converged = convergence_criteria.PreCriteria(computing_model_part, builder_and_solver.GetDofSet(), mechanical_solution_strategy.GetSystemMatrix(), mechanical_solution_strategy.GetSolutionVector(), mechanical_solution_strategy.GetSystemVector())
+            solver.SolveSolutionStep()
+            is_converged = convergence_criteria.PostCriteria(computing_model_part, builder_and_solver.GetDofSet(), mechanical_solution_strategy.GetSystemMatrix(), mechanical_solution_strategy.GetSolutionVector(), mechanical_solution_strategy.GetSystemVector())
+            if with_contact:
+                analysis._transfer_slave_to_master()
+            analysis.FinalizeSolutionStep()
+            if is_converged:
+                analysis.is_printing_rank = True
+                KratosMultiphysics.Logger.PrintInfo(analysis._GetSimulationName(), "Adaptative strategy converged in ", non_linear_iteration, "iterations" )
+                break
+            elif non_linear_iteration == analysis.non_linear_iterations:
+                analysis.is_printing_rank = True
+                KratosMultiphysics.Logger.PrintInfo(analysis._GetSimulationName(), "Adaptative strategy not converged after ", non_linear_iteration, "iterations" )
+                break
+            else:
+                if with_contact:
```

Comment:
probably this will be gone. But `with_contact` should not be in the core uitility.

Reply:
True, I will think on an alternative